manager/allocator/network.go

```diff
@@ -394,6 +383,80 @@ func (a *Allocator) getAllocatedNetworks() ([]*api.Network, error) {
 	return allocatedNetworks, nil
 }
 
+// getNodeNetworks returns all networks that should be allocated for a node
+func (a *Allocator) getNodeNetworks(nodeID string) ([]*api.Network, error) {
+	var (
+		// no need to initialize networks. we only append to it, and appending
+		// to a nil slice is valid. this has the added bonus of making this nil
+		// if we return an error
+		networks []*api.Network
+		err      error
+	)
+	a.store.View(func(tx store.ReadTx) {
+		// get all tasks currently assigned to this node. it's no big deal if
+		// the tasks change in the meantime, there's no race to clean up
+		// unneeded network attachments on a node.
+		var tasks []*api.Task
+		tasks, err = store.FindTasks(tx, store.ByNodeID(nodeID))
+		if err != nil {
+			return
+		}
+		// we need to keep track of network IDs that we've already added to the
+		// list of networks we're going to return. we could do
+		// map[string]*api.Network and then convert to []*api.Network and
+		// return that, but it seems cleaner to have a separate set and list.
+		networkIDs := map[string]struct{}{}
+		for _, task := range tasks {
+			// skip any tasks that are no longer relevant
+
+			// we only want tasks that are in desired state running. anything
+			// else is on its way out and irrelevant
```

Comment:
If a task is e.g. running but desired to be shut down, then presumably the load balancer IP is still in use, so I don't think it's safe to deallocate it yet.

e.g. if a node has two tasks

- Task1: state=running, desired=shutdown
- Task2: state=running, desired=shutdown

When Task2 terminates, we'll come here to decide whether we still need the load balancer IP. This function will say we can release it, I guess, even though we can't.

Reply:
I think my logic here was that this would only allocate new attachments, not free old ones, but you describe the actual case, and my logic was wrong. I will change it.